docs: Deep Links design document - application section (#1449)
Flank/flank-dashboard: metrics/web/docs/features/deep_links/01_deep_linking.md

Consider this class diagram that illustrates the required changes needed to parse the deep links from the URL:

![Parsing deep links diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/master/metrics/web/docs/features/deep_links/diagrams/parsing_deep_links_class_diagram.puml)

#### Applying Deep Links
The following subsections describe the changes related to the following requirements:
- restore the application state from deep links parameters;
- update deep links in response to application events.

##### PageParametersModel

> As we already know how the parse `query parameters`, we should introduce a new model that would represent the deserialized data from a query. For this purpose, we introduce a `PageParametersModel`.

Reviewer: 🏷️ 
```suggestion
As we already know how to parse `query parameters`, we should introduce a new model that would represent the deserialized data from a query. For this purpose, we introduce a `PageParametersModel`.
```

A `MetricsPage` should store a `RouteName` that was used to create this page and the `PageParametersModel` as an `arguments` field. The `MetricsPage` class should also provide an overridden `copyWith` method to copy a page with the updated parameters.

We should also update the `MetricsPageFactory` to create `MetricsPage`s with the given `RouteName` and `PageParametersModel`.

The code snippet below demonstrates the `MetricsPageFactory.create` changes:
```dart
  /// Creates the [MetricsPage] from the given [RouteConfiguration].
  MetricsPage create({
    RouteName routeName, 
    PageParametersModel pageParameters,
  }) {
    switch (routeName) {
      case RouteName.loading:
        return MetricsPage(
          child: const LoadingPage(),
          routeName: routeName,
          arguments: pageParameters,
        );
    /// Other MetricsPages creation...
```

##### NavigationNotifier
The `NavigationNotifier` is a class that holds the navigation logic of the Metrics Web application. The following subsections describe the main changes to the`NavigationNotifier` related to the `deep linking` feature implementation.

###### Updating page parameters

> When the `NavigationNotifier` receives a new `RouteConfiguration`, it should create a new `PageParametersModel` using a new `RouteConfiguration` instance and notify all listeners about the updates.
> 
> To do that, we should implement a new method `_updatePageParameters()` and call it whenever the current `RouteConfiguration` changes (in the `pop()` and the `_addNewPage()` methods).

Reviewer: 🏷️ 
```suggestion
When the `NavigationNotifier` receives a new `RouteConfiguration`, it should create a new `PageParametersModel` using a new `RouteConfiguration` instance and notify all listeners about the updates. For this, we should implement a new method `_updatePageParameters()` and call it whenever the current `RouteConfiguration` changes (in the `pop()` and the `_addNewPage()` methods).
```

> The following diagram demonstrates the general concept for updating the `PageParametersModel` when the current `RouteConfiguration` changes:
> ![Updating page parameters diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/deep_links_design_improvements/metrics/web/docs/features/deep_links/diagrams/updating_page_parameters_sequence_diagram.puml)

Reviewer: :readable:
This appears to be badly formatted. It seems like the empty line is missed: 
```suggestion
The following diagram demonstrates the general concept for updating the `PageParametersModel` when the current `RouteConfiguration` changes:

![Updating page parameters diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/deep_links_design_improvements/metrics/web/docs/features/deep_links/diagrams/updating_page_parameters_sequence_diagram.puml)
```

Now, when the `back button` is pressed, we should perform the following:
1. Pop the current page if the `NavigationNotifier.canPop` returns `true`;
2. If there is no underlying page, push and replace the current page with the `DashboardPage`.

Using such an approach allows users to navigate directly to the previous page if it exists, and to restore the applied `PageParametersModel` to the previous page if any.

##### Deep linking with authorization
Currently, the application pushes the `DashboardPage` when the user logs in, however, it would be much more user-friendly if the application redirects the user to the requested resource.

To improve this aspect of the navigation, we should modify the existing redirecting behaviour:
1. Use the `_redirectRoute` field to stores the redirect `RouteConfiguration` that requires authorization.
2. When the application pushes a new route that requires authorization, and the user is unauthenticated, set the `_redirectRoute`. If the route does not require authorization, clear the `_redirectRoute`.
3. When the user logs in, redirect the user to the `_redirectRoute` and clear the `_redirectRoute`.

> To implement such a behaviour, we should split the `.handleAuthenticationUpdates()` method of the `NavigationNotifier` into two different methods: `.handleLogOut()` and `.handleLogIn()`.

Reviewer: 🔖 
What if we name these methods as below? It seems like they actually handle the event (not log-in but logged-in):
```suggestion
To implement such a behaviour, we should split the `.handleAuthenticationUpdates()` method of the `NavigationNotifier` into two different methods: `.handleLoggedOut()` and `.handleLoggedIn()`.
```

Consider this class diagram that illustrates the required changes needed to parse the deep links from the URL:

![Parsing deep links diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/master/metrics/web/docs/features/deep_links/diagrams/parsing_deep_links_class_diagram.puml)

#### Applying Deep Links
The following subsections describe the required changes needed to be able to restore the application state from the deep links and update the deep links in response to application events.

##### PageParametersModel
Once we're able to parse `query parameters`, we should have a way to represent them in the Metrics Web application. To do that, let's introduce a `PageParametersModel`. 

> The `PageParametersModel` is an interface and does not store any data, however, it creates a general interface for all implementers including the `toMap() : Map<String, dynamic` method. 
> 
> The implementers of the `PageParametersModel` should include the parameters that are specific to a `page` they relate to, and implement the `PageParametersModel` interface.
> 
> The following class diagram demonstrates the general approach for creating new `PageParametersModel`s on a `SomePageParametersModel` example:

Reviewer: :readable:
It seems like an interface could not store any data itself.
```suggestion
The `PageParametersModel` is an interface for models that stores the application state parameters parsed from the query parameters. This interface provides the `.toMap()` method that the application uses to serialize appropriate models to the query parameters `Map`. The implementers of the `PageParametersModel` should include the parameters that are specific to a `page` they relate to. Also, they should know how to serialize and deserialize their data using the `.toMap()` method and `.fromMap()` constructor respectively. 

The following class diagram demonstrates the general approach for creating new `PageParametersModel`s on a `SomePageParametersModel` example:
```

Also, what do you think about renaming the `SomePageParametersModel` to `ConcretePageParametersModel` (to be more strict here) or `CoolPageParametersModel` (as in metrics web architecture documents).

![PageParametersModel class diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/deep_links_design_improvements/metrics/web/docs/features/deep_links/diagrams/page_parameters_model_class_diagram.puml)

##### PageParametersFactory

> Once we've implemented the required `PageParametersModel`, we should create a `PageParametersFactory` that encapsulates the creation of specific `PageParametersModel` from the given `RouteConfiguration`.

Reviewer: :readable:
```suggestion
The concrete `PageParametersModel` stores the data for concrete pages and knows how to convert this data from/into query parameters. However, the application should decide what implementation of `PageParametersModel` use for the current `RouteConfiguration`. The `PageParametersFactory` is to help here. This factory encapsulates the creation of specific `PageParametersModel` from the given `RouteConfiguration` calling the specific model deserialization.
```

Consider the following code snippet that demonstrates the `PageParametersFactory.create` method:
```dart
/// Returns the [PageParametersModel] based on the given [configuration].
PageParametersModel create(RouteConfiguration configuration) {
  final parameters = configuration.parameters;
  
  switch(configuration.name) {
    case RouteName.dashboard:
      return DashboardPageParameters.fromMap(parameters);
    case RouteName.projectGroups:

>       /// Other page parameters handling

Reviewer: 🏷️ 
Should it be a usual comment either a doc comment? 
```suggestion
      // Other page parameters handling
```

  }
}
```

> The `PageParametersFactory` should be injected into `NavigationNotifier`, since the `NavigationNotifier` holds all the navigation-related logic of the application and controls the deep linking of the application. Consider the [`NavigationNotifier`](#navigationnotifier) section describing this part of the application.

Reviewer: :label:
```suggestion
The `PageParametersFactory` should be injected into `NavigationNotifier` since the `NavigationNotifier` holds all the navigation-related logic of the application and controls its deep linking. Consider the [`NavigationNotifier`](#navigationnotifier) section to know more details in notifier-related changes.
```

##### MetricsPage and MetricsPageFactory

> To be able to restore the `PageParametersModel` when the application, for example, pops a page, we should update the `MetricsPage` and `MetricsPageFactory` classes.

Reviewer: :label:
```suggestion
To make the application able to restore the `PageParametersModel` in response to navigation events (e.g., popping a page), we should update the `MetricsPage` and `MetricsPageFactory` classes.
```

> A `MetricsPage` should store a `RouteName` used to create this page and the `PageParametersModel` as an `arguments` field. The `MetricsPage` class should also provide an overridden `copyWith` method to copy a page with the updated parameters.

Reviewer: 🏷️ 
```suggestion
A `MetricsPage` should store a `RouteName` that is [maybe "that was" is better] used to create this page and the `PageParametersModel` as an `arguments` field. The `MetricsPage` class should also provide an overridden `copyWith` method to copy a page with the updated parameters.
```

> The `MetricsPageFactory` should be updated as well to create `MetricsPage`s with the given `RouteName` and `PageParametersModel`.

Reviewer: :pencil:
It seems like we should prefer active voice where possible (https://developers.google.com/tech-writing/one/active-voice#prefer_active_voice_to_passive_voice).
```suggestion
We should also update the `MetricsPageFactory` to create `MetricsPage`s with the given `RouteName` and `PageParametersModel`.
```

The code snippet below demonstrates the `MetricsPageFactory.create` changes:
```dart
  /// Creates the [MetricsPage] from the given [RouteConfiguration].
  MetricsPage create({
    RouteName routeName, 
    PageParametersModel pageParameters,
  }) {
    switch (routeName) {
      case RouteName.loading:
        return MetricsPage(
          child: const LoadingPage(),
          routeName: routeName,
          arguments: pageParameters,
        );
    /// Other MetricsPages creation...
```

##### NavigationNotifier

> The `NavigationNotifier` is a class that holds the navigation logic of the Metrics Web application. The following subsections describe the main aspects needed to be able to implement the `deep linking` feature.

Reviewer: 🏷️ 
```suggestion
The `NavigationNotifier` is a class that holds the navigation logic of the Metrics Web application. The following subsections describe the main changes to the`NavigationNotifier` related to the `deep linking` feature implementation.
```

###### Updating page parameters

> When the `NavigationNotifier` gets a new `RouteConfiguration`, it should create a new `PageParametersModel` from the received `RouteConfiguration` and notify any listeners about the `PageParametersModel` updates.

Reviewer: :label:
Maybe "consumes" is a better word here? However, I'd like to suggest using the "receive" and replace its second appearance:
```suggestion
When the `NavigationNotifier` receives a new `RouteConfiguration`, it should create a new `PageParametersModel` using a new `RouteConfiguration` instance and notify all listeners about the updates.
```

This section describes the changes we should introduce to improve the overall navigation experience in the Metrics Web application.

##### Back Button navigation
Currently, the back button that is displayed on the `ProjectGroupPage` and on the `DebugMenuPage` navigates to the `DashboardPage` using the following code:
```dart
  /// Navigates to the [MetricsRoutes.dashboard] page.
void _navigateHome(BuildContext context) {
  final navigationNotifier = Provider.of<NavigationNotifier>(context, listen: false);

  navigationNotifier.push(MetricsRoutes.dashboard);
}
```

Formally, the current behaviour of the `Back Button` can be unexpected to the user. Consider the following examples describing why this behaviour may be unexpected:
- The user navigates to the `ProjectGroupPage` from the `DebugMenuPage`, and presses the `Back Button`. The application navigates the user to the `DashboardPage` instead of the `DebugMenuPage`;
- The user applies some filtering or sorting parameters on the `DashboardPage`, navigates to the `ProjectGroupPage`, and presses the `Back Button`. The application navigates the user to the `DashboardPage`, however, does not restore the page parameters.

> To improve the navigation experience here, we should introduce a new method to the `NavigationNotifier` - `canPop()`, that returns if the current page can be popped.

Reviewer: 🏷️ 
?
```suggestion
To improve the navigation experience here, we should introduce a new method to the `NavigationNotifier` - `canPop()`, that returns `true` if the current page can be popped.
```

Now, when the `back button` is pressed, we should perform the following:
1. Pop the current page if the `NavigationNotifier.canPop` returns `true`;
2. If there is no underlying page, push and replace the current page with the `DashboardPage`.

Using such an approach allows users to navigate directly to the previous page if it exists, and to restore the applied `PageParametersModel` to the previous page if any.

##### Deep linking with authorization
Currently, the application pushes the `DashboardPage` when the user logs in, however, it would be much more user-friendly if the application redirects the user to the requested resource.

To improve this aspect of the navigation, we should modify the existing redirecting behaviour:
1. Use the `_redirectRoute` field to stores the redirect `RouteConfiguration` that requires authorization.
2. When the application pushes a new route that requires authorization, and the user is unauthenticated, set the `_redirectRoute`. If the route does not require authorization, clear the `_redirectRoute`.
3. When the user logs in, redirect the user to the `_redirectRoute` and clear the `_redirectRoute`.

> To do that, we should split the `.handleAuthenticationUpdates()` method of the `NavigationNotifier` into 2 methods: `.handleLogOut()` and `.handleLogIn()`.

Reviewer: 🏷️ 
```suggestion
To implement such a bevavior, we should split the `.handleAuthenticationUpdates()` method of the `NavigationNotifier` into two different methods: `.handleLogOut()` and `.handleLogIn()`.
```
